test_rclcpp/test/parameter_fixtures.hpp

```diff
@@ -51,7 +51,7 @@ void declare_test_parameters(std::shared_ptr<rclcpp::Node> node, int declare_up_
   }
 
   for (int i = 0u; i < declare_up_to; ++i) {
-    node->declare_parameter(parameters[i].get_name());
+    node->declare_parameter(parameters[i].get_name(), parameters[i].get_parameter_value());
```

Comment:
What about declaring with the type instead?

```suggestion
    node->declare_parameter(parameters[i].get_name(), parameters[i].get_type());
```

Then I think there'd be less chance of interfering with the tests if they are also checking the value.

Reply:
When declaring only with the type, an user provided override is needed.
Between a default and an override there is no difference.

The other option is to modify the descriptor to make the parameter dynamically typed.

Reply:
oh, I thought that an override was only needed if the code tries to get the parameter before it is set. Is that not the case?

Reply:
I think we settled on throwing an exception in the declaration in this discussion https://github.com/ros2/rcl_interfaces/pull/118#discussion_r563976030

Reply:
You're right, I forgot about that compromise :grin: 

I think it's worth documenting the design decisions (e.g. why we went this this API), based on our prior discussions. Then we'll have something to reference when similar questions about this come up in the future. Perhaps as a quick note in a new `doc` directory in rcl_interfaces repo, or added to this design page: http://design.ros2.org/articles/ros_parameters.html

Reply:
Sounds good, I will open a new PR with docs